File: base/src/main/java/io/quarkus/code/model/ProjectDefinition.java
package io.quarkus.code.model;

import com.fasterxml.jackson.annotation.JsonCreator;
import com.fasterxml.jackson.annotation.JsonIgnore;
import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
import com.fasterxml.jackson.databind.annotation.JsonDeserialize;
import com.fasterxml.jackson.databind.annotation.JsonPOJOBuilder;
import jakarta.validation.constraints.NotEmpty;
import jakarta.validation.constraints.Pattern;

import java.util.List;
import java.util.Map;
import java.util.Objects;
import java.util.Set;

@JsonDeserialize(builder = ProjectDefinition.Builder.class)
@JsonIgnoreProperties(ignoreUnknown = true)
public record ProjectDefinition(
        String streamKey,

        @NotEmpty @Pattern(regexp = GROUPID_PATTERN) String groupId,

        @NotEmpty @Pattern(regexp = ARTIFACTID_PATTERN) String artifactId,

        @NotEmpty String version,

        @Pattern(regexp = CLASSNAME_PATTERN) String className,

        @Pattern(regexp = PATH_PATTERN) String path,

        @NotEmpty @Pattern(regexp = BUILDTOOL_PATTERN) String buildTool,
        Integer javaVersion,
        Boolean noCode,
        Boolean noExamples,
        Boolean noWrapper,
        Boolean noDockerfiles,
        Set<String> extensions,
        List<String> extraCodestarts,
        Map<String, String> codestartData) {

    public ProjectDefinition {
        Objects.requireNonNull(groupId, "groupId is required");
        Objects.requireNonNull(artifactId, "artifactId is required");
        Objects.requireNonNull(version, "version is required");
        Objects.requireNonNull(buildTool, "buildTool is required");
        Objects.requireNonNull(noCode, "noCode is required");
        Objects.requireNonNull(noExamples, "noExamples is required");
        Objects.requireNonNull(noWrapper, "noWrapper is required");
        Objects.requireNonNull(noDockerfiles, "noDockerfiles is required");
        Objects.requireNonNull(extensions, "extensions is required");
        Objects.requireNonNull(extraCodestarts, "extraCodestarts is required");
        Objects.requireNonNull(codestartData, "codestartData is required");
    }

    public static final String DEFAULT_GROUPID = "org.acme";
    public static final String DEFAULT_ARTIFACTID = "code-with-quarkus";
    public static final String DEFAULT_VERSION = "1.0.0-SNAPSHOT";
    public static final String DEFAULT_BUILDTOOL = "MAVEN";
    public static final Boolean DEFAULT_NO_CODE = false;
    public static final String DEFAULT_NO_CODE_STRING = "false";
    public static final Boolean DEFAULT_NO_WRAPPER = false;
    public static final String DEFAULT_NO_WRAPPER_STRING = "false";
    public static final Boolean DEFAULT_NO_DOCKERFILES = false;
    public static final String DEFAULT_NO_DOCKERFILES_STRING = "false";

    public static final String GROUPID_PATTERN = "^([a-zA-Z_$][a-zA-Z\\d_$]*\\.)*[a-zA-Z_$][a-zA-Z\\d_$]*$";
    public static final String ARTIFACTID_PATTERN = "^[a-z][a-z0-9-._]*$";
    public static final String CLASSNAME_PATTERN = GROUPID_PATTERN;
    public static final String PATH_PATTERN = "^\\/([a-z0-9\\-._~%!$&'()*+,;=:@]+\\/?)*$";
    public static final String BUILDTOOL_PATTERN = "^(MAVEN)|(GRADLE)|(GRADLE_KOTLIN_DSL)$";

    public static ProjectDefinition of() {
        return new ProjectDefinition(null, DEFAULT_GROUPID, DEFAULT_ARTIFACTID, DEFAULT_VERSION, null, null, DEFAULT_BUILDTOOL,
                null, DEFAULT_NO_CODE, DEFAULT_NO_CODE, DEFAULT_NO_WRAPPER, DEFAULT_NO_DOCKERFILES, Set.of(), List.of(),
                Map.of());
    }

    @JsonCreator
    public static Builder builder() {
        return new Builder();
    }

    @JsonPOJOBuilder(withPrefix = "")
    public static final class Builder {
        private String streamKey = null;
        private String groupId = DEFAULT_GROUPID;
        private String artifactId = DEFAULT_ARTIFACTID;
        private String version = DEFAULT_VERSION;
        private String className = null;
        private String path = null;
        private String buildTool = DEFAULT_BUILDTOOL;
        private Integer javaVersion = null;
        private Boolean noCode = DEFAULT_NO_CODE;
        private Boolean noExamples = DEFAULT_NO_CODE;
        private Boolean noWrapper = DEFAULT_NO_WRAPPER;
        private Boolean noDockerfiles = DEFAULT_NO_DOCKERFILES;
        private Set<String> extensions = Set.of();
        private List<String> extraCodestarts = List.of();
        private Map<String, String> codestartData = Map.of();

        private Builder() {
        }

        public Builder streamKey(String streamKey) {
            this.streamKey = streamKey;
            return this;
        }

        public Builder groupId(String groupId) {
            this.groupId = groupId;
            return this;
        }

        public Builder artifactId(String artifactId) {
            this.artifactId = artifactId;
            return this;
        }

        public Builder version(String version) {
            this.version = version;
            return this;
        }

        public Builder className(String className) {
            this.className = className;
            return this;
        }

        public Builder path(String path) {
            this.path = path;
            return this;
        }

        public Builder buildTool(String buildTool) {
            this.buildTool = buildTool;
            return this;
        }

        public Builder javaVersion(Integer javaVersion) {
            this.javaVersion = javaVersion;
            return this;
        }

        public Builder noCode(Boolean noCode) {
            this.noCode = noCode;
            return this;
        }

        public Builder noExamples(Boolean noExamples) {
            this.noExamples = noExamples;
            return this;
        }

        public Builder noWrapper(Boolean noWrapper) {
            this.noWrapper = noWrapper;
            return this;
        }

        public Builder noDockerfiles(Boolean noDockerfiles) {
            this.noDockerfiles = noDockerfiles;
            return this;
        }

        public Builder extensions(Set<String> extensions) {
            this.extensions = extensions;
            return this;
        }

        public Builder extraCodestarts(List<String> extraCodestarts) {
            this.extraCodestarts = extraCodestarts;
            return this;
        }

        public Builder codestartData(Map<String, String> codestartData) {
            this.codestartData = codestartData;
            return this;
        }

        public ProjectDefinition build() {
            return new ProjectDefinition(streamKey, groupId, artifactId, version, className, path, buildTool, javaVersion,
                    noCode, noExamples, noWrapper, noDockerfiles, extensions, extraCodestarts, codestartData);
        }
    }

}

File: src/main/kotlin/io/quarkus/code/model/ProjectDefinition.kt
package io.quarkus.code.model

import com.fasterxml.jackson.annotation.JsonIgnoreProperties
import org.eclipse.microprofile.openapi.annotations.parameters.Parameter
import javax.validation.constraints.NotEmpty
import javax.validation.constraints.Pattern
import javax.ws.rs.DefaultValue
import javax.ws.rs.QueryParam

@JsonIgnoreProperties(ignoreUnknown = true)
class ProjectDefinition {

    companion object {
        const val DEFAULT_GROUPID = "org.acme"
        const val DEFAULT_ARTIFACTID = "code-with-quarkus"
        const val DEFAULT_VERSION = "1.0.0-SNAPSHOT"
        const val DEFAULT_BUILDTOOL = "MAVEN"
        const val DEFAULT_NO_CODE = false

        const val GROUPID_PATTERN = "^([a-zA-Z_\$][a-zA-Z\\d_\$]*\\.)*[a-zA-Z_\$][a-zA-Z\\d_\$]*\$"
        const val ARTIFACTID_PATTERN = "^[a-z][a-z0-9-._]*\$"
        const val CLASSNAME_PATTERN = GROUPID_PATTERN
        const val PATH_PATTERN = "^\\/([a-z0-9\\-._~%!\$&'()*+,;=:@]+\\/?)*\$"
        const val BUILDTOOL_PATTERN = "^(MAVEN)|(GRADLE)|(GRADLE_KOTLIN_DSL)\$"
    }

    constructor()
    constructor(streamKey: String? = null,
                groupId: String = DEFAULT_GROUPID,
                artifactId: String = DEFAULT_ARTIFACTID,
                version: String = DEFAULT_VERSION,
                className: String? = null,
                path: String? = null,
                buildTool: String = DEFAULT_BUILDTOOL,
                noExamples: Boolean = false,
                noCode: Boolean = false,
                extensions: Set<String> = setOf(),
                shortExtensions: String = "") {
        this.streamKey = streamKey
        this.groupId = groupId
        this.artifactId = artifactId
        this.version = version
        this.className = className
        this.extensions = extensions
        this.path = path
        this.buildTool = buildTool
        this.noCode = noCode
        this.noExamples = noExamples
        this.shortExtensions = shortExtensions
    }

    @QueryParam("S")
    @Parameter(name = "S", description = "The platform stream to use to create this project ('platformKey:streamId' or 'streamId')", required = false)
    var streamKey: String? = null
        private set

    @DefaultValue(DEFAULT_GROUPID)
    @NotEmpty
    @Pattern(regexp = GROUPID_PATTERN)
    @QueryParam("g")
    @Parameter(name = "g", description = "GAV: groupId", required = false)
    var groupId: String = DEFAULT_GROUPID
        private set

    @DefaultValue(DEFAULT_ARTIFACTID)
    @NotEmpty
    @Pattern(regexp = ARTIFACTID_PATTERN)
    @QueryParam("a")
    @Parameter(name = "a", description = "GAV: artifactId", required = false)
    var artifactId: String = DEFAULT_ARTIFACTID
        private set

    @DefaultValue(DEFAULT_VERSION)
    @NotEmpty
    @QueryParam("v")
    @Parameter(name = "v", description = "GAV: version", required = false)
    var version: String = DEFAULT_VERSION
        private set

    @QueryParam("c")
    @Pattern(regexp = CLASSNAME_PATTERN)
    @Parameter(name = "c", description = "The class name to use in the generated application", required = false)
    var className: String? = null
        private set

    @QueryParam("p")
    @Pattern(regexp = PATH_PATTERN)
    @Parameter(name = "p", description = "The path of the REST endpoint created in the generated application", required = false)
    var path: String? = null
        private set

    @DefaultValue(DEFAULT_NO_CODE.toString())
    @QueryParam("ne")
    @Parameter(name = "ne", description = "No code examples (Deprecated: use noCode (nc) instead)", required = false)
    @Deprecated(message = "Use noCode (nc) instead")
    var noExamples: Boolean = DEFAULT_NO_CODE
        private set

    @DefaultValue(DEFAULT_NO_CODE.toString())
    @QueryParam("nc")
    @Parameter(name = "nc", description = "No code", required = false)
        var noCode: Boolean = DEFAULT_NO_CODE
        private set

    @DefaultValue(DEFAULT_BUILDTOOL)
    @NotEmpty
    @QueryParam("b")
    @Pattern(regexp = BUILDTOOL_PATTERN)
    @Parameter(name = "b", description = "The build tool to use (MAVEN or GRADLE)", required = false)
    var buildTool: String = DEFAULT_BUILDTOOL
        private set

    @QueryParam("e")
    @Parameter(name = "e", description = "The set of extension ids that will be included in the generated application", required = false)
    var extensions: Set<String> = setOf()
        private set

    @QueryParam("s")
    @DefaultValue("")
    @Parameter(name = "s", description = "The set of extension shortIds separated by a '.' that will be included in the generated application (Deprecated: see https://github.com/quarkusio/code.quarkus.io/issues/424)", required = false)
    @Deprecated(message = "see https://github.com/quarkusio/code.quarkus.io/issues/424")
    var shortExtensions: String = ""
        private set

    override fun toString(): String {
        return "QuarkusProject(streamKey='$streamKey', groupId='$groupId', artifactId='$artifactId', version='$version', className='$className', path='$path', buildTool='$buildTool', noCode='$noCode', extensions=$extensions')"
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false

        other as ProjectDefinition

        if (streamKey != other.streamKey) return false
        if (groupId != other.groupId) return false
        if (artifactId != other.artifactId) return false
        if (version != other.version) return false
        if (className != other.className) return false
        if (path != other.path) return false
        if (noExamples != other.noExamples) return false
        if (noCode != other.noCode) return false
        if (buildTool != other.buildTool) return false
        if (extensions != other.extensions) return false
        if (shortExtensions != other.shortExtensions) return false

        return true
    }

    override fun hashCode(): Int {
        var result = streamKey?.hashCode() ?: 0
        result = 31 * result + groupId.hashCode()
        result = 31 * result + artifactId.hashCode()
        result = 31 * result + version.hashCode()
        result = 31 * result + (className?.hashCode() ?: 0)
        result = 31 * result + (path?.hashCode() ?: 0)
        result = 31 * result + noExamples.hashCode()
        result = 31 * result + noCode.hashCode()
        result = 31 * result + buildTool.hashCode()
        result = 31 * result + extensions.hashCode()
        result = 31 * result + shortExtensions.hashCode()
        return result
    }
}